flax/nnx/nn/normalization.py

```diff
@@ -988,3 +1003,209 @@ def __call__(self, x, *, mask: tp.Optional[jax.Array] = None):
+        batch_stats[new_path] = bstat
+
+    self.batch_stats = nnx.data(batch_stats)
+
+  def __call__(
+    self,
+    x,
+    update_stats: tp.Optional[bool] = None,
+  ):
+    """Compute the largest singular value of the weights in ``self.layer_instance``
+    using power iteration and normalize the weights using this value before
+    computing the ``__call__`` output.
+
+    Args:
+      x: the input array of the nested layer
+      update_stats: if True, update the internal ``u`` vector and ``sigma``
+        value after computing their updated values using power iteration. This
+        will help the power iteration method approximate the true singular value
+        more accurately over time.
+
+    Returns:
+      Output of the layer using spectral normalized weights.
+    """
+    update_stats = first_from(
+      update_stats,
+      not self.use_running_average,
+      error_msg="""No `update_stats` argument was provided to SpectralNorm
+        as either a __call__ argument or __init__ argument.""",
+    )
+    state = nnx.state(self.layer_instance, nnx.Param)
+    for path, param in nnx.to_flat_state(state):
+      self._spectral_normalize_inplace(path, param, update_stats=update_stats)
+    return self.layer_instance(x)
+
+  def _spectral_normalize_inplace(self, path, orig_param, update_stats):
+    param = orig_param
+    param_shape = param.shape
+    if param.ndim <= 1 or self.n_steps < 1:
+      return
+    elif param.ndim > 2:
+      if self.error_on_non_matrix:
+        raise ValueError(
+          f'Layer instance parameter is {param.ndim}D but error_on_non_matrix is True'
+        )
+      else:
+        param = jnp.reshape(param, (-1, param.shape[-1]))
+
+    path_u = path + ("u", )
+    path_sigma = path + ("sigma", )
+
+    if path_u not in self.batch_stats:
+      raise RuntimeError(
+        f"Could not find the path for u batch stat corresponding to the param {path} "
+        "in the batch stats dict. Parameters of the layer_instance should not change!"
+      )
+    if path_sigma not in self.batch_stats:
+      raise RuntimeError(
+        f"Could not find the path for sigma batch stat corresponding to the param {path} "
+        "in the batch stats dict. Parameters of the layer_instance should not change!"
+      )
+
+    u = self.batch_stats[path_u].value
+
+    for _ in range(self.n_steps):
+      v = _l2_normalize(jnp.matmul(u, param.T), eps=self.epsilon)
+      u = _l2_normalize(jnp.matmul(v, param), eps=self.epsilon)
+
+    u = lax.stop_gradient(u)
+    v = lax.stop_gradient(v)
+
+    sigma = jnp.matmul(jnp.matmul(v, param), u.T)[0, 0]
+    param /= jnp.where(sigma != 0, sigma, 1)
+    param = param.reshape(param_shape)
+
+    if update_stats:
+      self.batch_stats[path_u][...] = u
+      self.batch_stats[path_sigma][...] = sigma
+
+    dtype = dtypes.canonicalize_dtype(param, u, v, sigma, dtype=self.dtype)
+    orig_param[...] = jnp.asarray(param, dtype)
```

Comment:
@cgarciae Here we should use `.value`, right?
```suggestion
    orig_param.value = jnp.asarray(param, dtype)
```